thinkhazard/processing/harvesting.py

```diff
@@ -394,6 +398,26 @@ def harvest_layers(self, hazard_type=None):
                 )
                 logger.error(traceback.format_exc())
 
+        layers_db_end = self.dbsession.query(Layer).all()
+        for layer in layers_db_end:
+            if layer.is_harvested() is False:
+                logger.info("Deleting layer {}".format(layer.hazardset_id))
+                self.dbsession.delete(layer)
+
+        hazardsets = self.dbsession.query(HazardSet).all()
+        for hazardset in hazardsets:
+            if len(hazardset.layers) == 0:
+                logger.info("Deleting hazardset {} with output".format(hazardset.id))
+                self.dbsession.query(Output) \
+                    .filter(Output.hazardset_id == hazardset.id) \
+                    .delete()
+                self.dbsession.flush()
+                self.dbsession.delete(hazardset)
+            else:
+                logger.info("Keeping hazardset {}".format(hazardset.id))
+                hazardset.completed = False
+                hazardset.processed = None
```

Comment:
You should not iterate on hazardsets, but just alter hazardsets for not harvested layers.
Here you invalidate all processing outputs even if nothing has changed.
Here is the example I gave you in slack:
```
if layer.is_harvested() is False:
    session.delete(layer)
    hazardset = layer.hazardset
    session.query(Output).filter(Output.hazardset_id == hazardset.id).delete()
    session.flush()
    if len(hazardset.layers) == 0:
        session.delete(hazardset)
    else
        hazardset.completed = False
        hazardset.processed = None
```

Reply:
Ok thanks I moved it back into the layer iteration as discussed (just couldn't get pop to work). However the output deletion still doesn't seem correct.